design/seed_vm.md

```diff
@@ -0,0 +1,190 @@
+# SeedVM
+
+## Introduction
+
+SeedVM is a register based virtual machine which can interpret and execute the
+SeedLang byte code. Register based virtual machine avoids several push and pop
+instructions that stack based virtual machine needs to move values around the
+stack. These instructions are particularly expensive, because they involve copy
+of the values.
+
+The design intention of SeedVM is to make a stable, secure and fast virtual
+machine for the visualizable low-code programming environment that focuses on
+educational purposes.
+
+The SeedLang program is translated from SeedAST to byte code by the compiler.
+The design of the byte code is derived from the Lua 5.1. Please refer to [A
+No-Frills Introduction to Lua 5.1 VM
+Instructions](http://underpop.free.fr/l/lua/docs/a-no-frills-introduction-to-lua-5.1-vm-instructions.pdf).
+
+## SeedLang Value Types
+
+SeedLang is a dynamically typed programming language. Types are attached to
+values rather then variables. There are 9 types in SeedLang: Nil, Boolean,
+Number, String, List, Table, Function, Module and SeedObject:
+
+- Nil: Nil value, only one value nil
+- Boolean: Boolean value, true or false
+- Number: Double precision floating point number
+- String: A sequential collection of characters to represent a text
+- List: A list of values that can be access by index
+- Table: A dictionary to hold key and value pairs
+- Function: Function object which includes module level function, member
+  function or build-in function
+- Module: Module object which can hold functions and objects in it
+- SeedObject: The base type of all the customer objects
```

Comment:
I decide to remove SeedObject. My intention of adding SeedObject is to implement following SeedPython code:
```python
obj = SeedObject({
    'x': 10,
    'y': 20,
    'add': m_add,
    'inner_obj': {
        'x': 100,
        'y': 200,
        'add': m_add,
    },
})
```
But it's easy for compiler to translate this code into a Table object. As we only support object now. the name of obj is enough to identify the object.
We may need a SeedObject base class for all the SeedAST or SeedVM values. But this might be a host language class and implement detail, no need to state here.